lib/symbols.nix

```diff
@@ -0,0 +1,29 @@
+rec {
+  /* mkSymbol returns a value that is only comparable to itself. The main
+     use is to create unique identifiers or labels and enables a form of weak
+     encapsulation, or information hiding.
+
+     The reason this works is that function equality works by comparing the
+     memory location. So two mkSymbol calls with the same description will
+     generate two different singletons.
```

Comment:
that's not actually true, functions are compare unequal. what you're (ab)using here is some legacy support behaviour: https://github.com/NixOS/nix/blob/master/src/libexpr/eval.cc#L2350

Reply:
That was the whole point of the symbol, so it cannot clash with another code base that defines a similarly-named symbol.

It's a bit like complaining that two different Haskell Enum symbols defined in different modules aren't inter-operable.

Reply:
At least Haskell will tell you what the problem is, whereas `lib.symbols` can't tell you that you're using a symbol from the wrong file. It just fails silently.

Perhaps `lib.symbols` could use the same pattern to distinguish evaluations of `lib`, so that comparisons of symbols from one `lib` don't return `false` but throw an error. This way, the user will have some clue as to what went wrong.
This only truly solves this error for symbols that are put into `lib` itself, but may catch some errors with non-`lib` symbols as well, if they happen to be constructed using different `lib`s.

Even with this improvement, I'm not convinced.